src/hotspot/share/gc/shared/pretouchTask.cpp

```diff
@@ -28,59 +28,99 @@
 #include "runtime/atomic.hpp"
 #include "runtime/globals.hpp"
 #include "runtime/os.hpp"
+#include "utilities/ticks.hpp"
 
 PretouchTask::PretouchTask(const char* task_name,
                            char* start_address,
                            char* end_address,
                            size_t page_size,
-                           size_t chunk_size) :
+                           size_t chunk_size,
+                           uint n_threads,
+                           size_t task_status) :
     AbstractGangTask(task_name),
     _cur_addr(start_address),
     _start_addr(start_address),
     _end_addr(end_address),
     _page_size(page_size),
-    _chunk_size(chunk_size) {
+    _chunk_size(chunk_size),
+    _n_threads(n_threads),
+    _task_status(task_status) {
 
   assert(chunk_size >= page_size,
          "Chunk size " SIZE_FORMAT " is smaller than page size " SIZE_FORMAT,
          chunk_size, page_size);
 }
 
+void PretouchTask::reinitialize(char* start_addr, char* end_addr) {
+  Atomic::release_store(&_cur_addr, start_addr);
+  Atomic::release_store(&_end_addr, end_addr);
+}
+
 size_t PretouchTask::chunk_size() {
   return PreTouchParallelChunkSize;
 }
 
 void PretouchTask::work(uint worker_id) {
+
+  // Following atomic loads are required to make other processor store
+  // visible to all threads from this points.
+  char *cur_addr = Atomic::load(&_cur_addr);
+  char *end_addr = Atomic::load(&_end_addr);
+  OrderAccess::fence();
+
+  // Required to avoid un-necessary update of _cur_addr once the task is done.
+  if ( cur_addr >= end_addr ) {
+    return ;
+  }
+
+  uint thread_num = Atomic::add(&_n_threads, 1u);
+
   while (true) {
     char* touch_addr = Atomic::fetch_and_add(&_cur_addr, _chunk_size);
     if (touch_addr < _start_addr || touch_addr >= _end_addr) {
       break;
     }
 
-    char* end_addr = touch_addr + MIN2(_chunk_size, pointer_delta(_end_addr, touch_addr, sizeof(char)));
+    end_addr = touch_addr + MIN2(_chunk_size, pointer_delta(_end_addr, touch_addr, sizeof(char)));
 
     os::pretouch_memory(touch_addr, end_addr, _page_size);
+
+  }
+
+  // Mark task done only when the last thread finishes its work.
+  thread_num = Atomic::sub(&_n_threads, 1u);
+
+  if (thread_num == 0) {
+    Atomic::release_store(&_cur_addr, _end_addr);
```

Comment:
The release is unnecessary given the Atomic::sub.

Reply:
Besides what David said, I think this is also confusing.  I think it's to account for the claim (using fetch_and_add) unconditionally adding the chunk size to cur_addr, which could lead to overshoot.  I think it would be clearer to prevent the overshoot from occurring.

Reply:
David, will update as suggested.

Kim, thanks for your feedback. Updating **_cur_addr** was not needed here but thought to keep the markers in clean state. Again, **_cur_addr** & **_end_addr** gets updated before the start of next pretouch resize so will remove this. If something breaks then will fix as suggested. I hope this should be OK.